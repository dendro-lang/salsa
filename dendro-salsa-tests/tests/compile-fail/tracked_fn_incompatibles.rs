#[salsa::jar(db = Db)]
struct Jar(MyInput, tracked_fn_with_data, tracked_fn_with_db, tracked_fn_with_constructor, tracked_fn_with_one_input, tracked_fn_with_receiver_not_applied_to_impl_block);

trait Db: salsa::DbWithJar<Jar> {}

#[salsa::input(jar = Jar)]
struct MyInput {
    field: u32,
}


#[salsa::tracked(jar = Jar, data = Data)]
fn tracked_fn_with_data(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}

#[salsa::tracked(jar = Jar, db = Db)]
fn tracked_fn_with_db(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}

#[salsa::tracked(jar = Jar, constructor = TrackedFn3)]
fn tracked_fn_with_constructor(db: &dyn Db, input: MyInput) -> u32 {
    input.field(db) * 2
}


#[salsa::tracked(jar = Jar)]
fn tracked_fn_with_one_input(db: &dyn Db) -> u32 {
}


#[salsa::tracked(jar = Jar)]
fn tracked_fn_with_receiver_not_applied_to_impl_block(&self, db: &dyn Db) -> u32 {
}

#[salsa::tracked(jar = Jar, specify)]
fn tracked_fn_with_receiver_not_applied_to_impl_block(db: &dyn Db, input: MyInput, input: MyInput) -> u32 {
}







fn main() {}