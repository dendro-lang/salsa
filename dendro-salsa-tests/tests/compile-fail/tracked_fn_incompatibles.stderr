error: `data` option not allowed here
  --> tests/compile-fail/tracked_fn_incompatibles.rs:12:29
   |
12 | #[salsa::tracked(jar = Jar, data = Data)]
   |                             ^^^^

error: `db` option not allowed here
  --> tests/compile-fail/tracked_fn_incompatibles.rs:17:29
   |
17 | #[salsa::tracked(jar = Jar, db = Db)]
   |                             ^^

error: `constructor` option not allowed here
  --> tests/compile-fail/tracked_fn_incompatibles.rs:22:29
   |
22 | #[salsa::tracked(jar = Jar, constructor = TrackedFn3)]
   |                             ^^^^^^^^^^^

error: #[salsa::tracked] must also be applied to the impl block for tracked methods
  --> tests/compile-fail/tracked_fn_incompatibles.rs:34:55
   |
34 | fn tracked_fn_with_receiver_not_applied_to_impl_block(&self, db: &dyn Db) -> u32 {
   |                                                       ^

error: tracked function takes too many arguments to have its value set with `specify`
  --> tests/compile-fail/tracked_fn_incompatibles.rs:37:29
   |
37 | #[salsa::tracked(jar = Jar, specify)]
   |                             ^^^^^^^

error[E0412]: cannot find type `tracked_fn_with_data` in this scope
 --> tests/compile-fail/tracked_fn_incompatibles.rs:2:21
  |
2 | struct Jar(MyInput, tracked_fn_with_data, tracked_fn_with_db, tracked_fn_with_constructor, tracked_fn_with_one_input, tracked_fn_with_rec...
  |                     ^^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `tracked_fn_with_db` in this scope
 --> tests/compile-fail/tracked_fn_incompatibles.rs:2:43
  |
2 | struct Jar(MyInput, tracked_fn_with_data, tracked_fn_with_db, tracked_fn_with_constructor, tracked_fn_with_one_input, tracked_fn_with_rec...
  |                                           ^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0412]: cannot find type `tracked_fn_with_constructor` in this scope
  --> tests/compile-fail/tracked_fn_incompatibles.rs:2:63
   |
2  | struct Jar(MyInput, tracked_fn_with_data, tracked_fn_with_db, tracked_fn_with_constructor, tracked_fn_with_one_input, tracked_fn_with_rec...
   |                                                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: a struct with a similar name exists: `tracked_fn_with_one_input`
...
28 | #[salsa::tracked(jar = Jar)]
   | ---------------------------- similarly named struct `tracked_fn_with_one_input` defined here

error[E0412]: cannot find type `tracked_fn_with_receiver_not_applied_to_impl_block` in this scope
 --> tests/compile-fail/tracked_fn_incompatibles.rs:2:119
  |
2 | ...r, tracked_fn_with_one_input, tracked_fn_with_receiver_not_applied_to_impl_block);
  |                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not found in this scope

error[E0308]: mismatched types
  --> tests/compile-fail/tracked_fn_incompatibles.rs:29:46
   |
29 | fn tracked_fn_with_one_input(db: &dyn Db) -> u32 {
   |    -------------------------                 ^^^ expected `u32`, found `()`
   |    |
   |    implicitly returns `()` as its body has no tail or `return` expression
